train_redpajama.py

```diff
@@ -0,0 +1,312 @@
+import os
+import math
+import glob
+import time
+from functools import partial
+from pathlib import Path
+from typing import Tuple, Optional
+
+import lightning as L
+from lightning.fabric.strategies import FSDPStrategy
+
+import torch
+from torch.utils.data import DataLoader
+from torch.distributed.fsdp.wrap import transformer_auto_wrap_policy
+
+import numpy as np
+
+from lit_stablelm.model import Block, StableLM, StableLMConfig
+from lit_stablelm.packed_dataset import PackedDataset, CombinedDataset
+from lit_stablelm.utils import save_model_checkpoint
+
+
+out_dir = "out/training"
+save_interval = 1000
+eval_interval = 1000
+eval_iters = 100
+log_interval = 1
+
+# compile = False
+
+# Hyperparameters
+learning_rate = 6e-4
+batch_size = 125
+micro_batch_size = 5
+max_iters = 600000  # num_epochs * epoch_size // devices
+weight_decay = 1e-1
+beta1 = 0.9
+beta2 = 0.95
+grad_clip = 1.0
+decay_lr = True
+warmup_iters = 2000
+lr_decay_iters = max_iters
+min_lr = 6e-5
+
+
+# Data proportions from https://arxiv.org/pdf/2302.13971.pdf Table 1
+data_config = [
+    ("arxiv", 2.5),
+    ("book", 4.5),
+    ("c4", 15.0),
+    ("cc", 67.0),
+    ("github", 4.5),
+    ("stackexchange", 2.0),
+    ("wikipedia", 4.5),
+]
+
+
+def main(
+    devices: int = 4,
+    train_data_dir: Path = "data/lit-redpajama",
+    val_data_dir: Optional[Path] = None,
+) -> None:
+    auto_wrap_policy = partial(
+        transformer_auto_wrap_policy, transformer_layer_cls={Block}
+    )
+    strategy = FSDPStrategy(
+        auto_wrap_policy=auto_wrap_policy, activation_checkpointing=Block
+    )
+
+    fabric = L.Fabric(
+        accelerator="cuda", devices=devices, precision="bf16-mixed", strategy=strategy
+    )
+    fabric.launch()
+    fabric.seed_everything(1337)
+
+    if fabric.global_rank == 0:
+        os.makedirs(out_dir, exist_ok=True)
+
+    config = StableLMConfig.from_name("7B")
+
+    train_dataloader, val_dataloader = create_dataloaders(
+        batch_size=micro_batch_size,
+        block_size=config.block_size,
+        fabric=fabric,
+        train_data_dir=train_data_dir,
+        val_data_dir=val_data_dir,
+        seed=1338,
+    )
+    train_dataloader, val_dataloader = fabric.setup_dataloaders(train_dataloader, val_dataloader)
```

Comment:
This needs https://github.com/Lightning-AI/lit-llama/commit/b3a24835ec6b4882963a04ecdb3d01176d23c533

Reply:
we're running on cutting edge anyway right?

Reply:
Fabric doesn't solve this issue. The problem is that the val_data_dir is None